engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala

```diff
@@ -371,12 +403,18 @@ object ChangesetBuilder {
     * @param edit the location of the edit
     * @return the invalidated nodes of the tree
     */
-  private def invalidated(tree: Tree, edit: Location): Tree = {
+  private def invalidated(
+    tree: Tree,
+    edit: Location,
+    onlyLeafs: Boolean
```

Comment:
and why do we need this parameter now?

Reply:
Without f867313 the _"should module with undefined literal"_ test fails:
```
$ git show f867313 | patch -p1 -R
$ sbt runtime/testOnly *ChangesetBuilderTest
[info] - should module with undefined literal *** FAILED ***
[info]   Set(b1c393b2-67be-488b-b46d-2adba21bca6d, c05a82f3-9dea-49fb-81ce-8dfa5703e5ac)
 did not contain the same elements as
 List(c05a82f3-9dea-49fb-81ce-8dfa5703e5ac) (ChangesetBuilderTest.scala:368)
```
The additional `b1c393b2-67be-488b-b46d-2adba21bca6d` id represents `1 + undefined` text. The `fillInHoles` function associates the space between `+` and `undefined` to `b1c393b2-67be-488b-b46d-2adba21bca6d` - as the edit replaces `undefined` with `42` it touches the space on the boundary and as such we get two change sets. To keep the original behavior we first work on _"leafs"_ only as before and only when they find nothing, we try harder with non-leaf elements.